File: src/NotificationPortal.Web/Controllers/ChallengeController.cs
using Microsoft.AspNetCore.Mvc;
using NotificationPortal.Web.ActorModel;
using NotificationPortal.Web.Models;

namespace NotificationPortal.Web.Controllers
{
    public class ChallengeController : Controller
    {
        private readonly RelogifyActorModel _actorModel;

        public ChallengeController(RelogifyActorModel actorModel)
        {
            _actorModel = actorModel;
        }

        [HttpGet]
        public IActionResult SendChallenge()
        {
            return View(
                new SendChallengeModel(CommunityName: "", FromPlayer: "", ToPlayer: "", RequestStatusMessage: null));
        }

        [HttpGet]
        public IActionResult ChallengeList()
        {
            return View();
        }

        [HttpPost]
        [ValidateAntiForgeryToken]
        public IActionResult SendChallenge(SendChallengeModel model)
        {
            if (!ModelState.IsValid)
                return View(model);

            _actorModel.PublishMessage(new ChallengeIssuedMessage(SendChallengeModel: model));

            return View(model with { RequestStatusMessage = "Challenge queued for sending" });
        }
    }
}

File: src/NotificationPortal.Web/Controllers/ChallengeController.fs
namespace NotificationPortal.Web.Controllers

open Microsoft.AspNetCore.Mvc
open NotificationPortal.Data
open System.ComponentModel.DataAnnotations

[<CLIMutable>]
type SendChallengeModel = {
    [<Required>]
    CommunityName: string

    [<Required>]
    FromPlayer: string

    [<Required>]
    ToPlayer: string

    RequestStatusMessage: string
}

type ChallengeController () =
    inherit Controller()

    [<HttpGet>]
    member this.SendChallenge () =
        let emptySendChallengeModel: SendChallengeModel =
            { CommunityName = ""
              FromPlayer = ""
              ToPlayer = ""
              RequestStatusMessage = "" }

        this.View(emptySendChallengeModel)

    [<HttpGet>]
    member this.ChallengeList () =
        this.View()

    [<HttpPost>]
    [<ValidateAntiForgeryToken>]
    member this.SendChallenge (model: SendChallengeModel) =
        if not this.ModelState.IsValid then
            this.View(model)
        else
            // TODO: Call challengeService, try/catch, return success or error
            this.View({ model with RequestStatusMessage = "Challenge queued for sending" })


